### PR TITLE
Add checkpoint UI

### PR DESCRIPTION
**Pull request by Karakatiza666**
Related PR: https://github.com/feldera/feldera/pull/6003

[Screencast from 2026-05-07 03-05-38.webm](https://github.com/user-attachments/assets/3415c680-5b25-4ce7-8a75-489103d67e35)

image

Removed  API-initiated transaction phase status information because it is generally not needed

### CODE REVIEW

File: js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte [hunk 1]
```diff
@@ -139,6 +154,28 @@
...
+  $effect(() => {
+    pipelineName
+    if (!metricsAvailable) {
+      checkpoints = []
+      checkpointStatus = null
+      return
+    }
+    // Poll checkpoint-related endpoints so the UI stays current with
+    // ongoing checkpoint activity (also needed for the mock simulator).
+    const fetchCheckpoints = () => {
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte:167]: This PR adds a substantial new web-console behavior surface: checkpoint polling, new checkpoint status/indicator UI, drawer state changes, and new manager calls. Per our standing web-console rule, that needs frontend tests before approval. A good minimal bar here would be Vitest/@testing-library coverage that exercises the checkpoint activity rendering and the new checkpoints drawer / request-checkpoint flow, with the API calls mocked via `usePipelineManager`.

**Review by mythical-fred — CHANGES_REQUESTED** [review]
This web-console behavior change still needs frontend coverage; see inline.

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      >
+      <button class="btn preset-filled-surface-100-900" onclick={onShowCheckpoints}>
+        All checkpoints
+      </button>
+    </div>
+  {/if}
+  <div class="hr"></div>
+{/snippet}
+
+{#if activityVisible || last || true}
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte]: `{#if activityVisible || last || true}` — the `|| true` makes this guard always true, so the wrapper renders (with its `min-h-16` and the leading `<div class="hr">` snippet) even when none of the inner branches match. Looks like a debug leftover; drop the `|| true` (and probably also the `|| true` makes both `activityVisible` and `last` redundant in the guard, so the intent of the original condition is unclear — please make it explicit).

File: js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte [hunk 3]
```diff
@@ -15,7 +15,9 @@
   import ConnectorErrors, {
     type ConnectorErrorFilter
   } from '$lib/components/pipelines/editor/performance/ConnectorErrors.svelte'
+  import CheckpointsStatus from '$lib/components/pipelines/editor/performance/CheckpointsStatus.svelte'
   import { useIsScreenXl } from '$lib/compositions/layout/useIsMobile.svelte'
+  import { usePremiumFeatures } from '$lib/compositions/usePremiumFeatures.svelte'
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/components/pipelines/editor/TabPerformance.svelte]: `usePremiumFeatures` is imported here but never called in this file (the gate lives inside `CheckpointsIndicator.svelte`). Remove the unused import.

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/TransactionStatus.svelte [hunk 4]
```diff
@@ -2,17 +2,17 @@
...
 
-  let { metrics }: { metrics: { current: PipelineMetrics } } = $props()
+  let { metrics, class: _class = '' }: { metrics: { current: PipelineMetrics }; class?: string } =
+    $props()
 
   const global = $derived(metrics.current.global)
   const transactionStatus = $derived(global.transaction_status)
   const commitProgress = $derived(global.commit_progress)
-  const initiatedByApi = $derived(global.transaction_initiators?.initiated_by_api)
+  // const apiInitiatedPhase = $derived(global.transaction_initiators?.initiated_by_api ?? null)
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/TransactionStatus.svelte]: Don't commit commented-out code. Either keep API-initiated phase as a feature (it was visible in the previous UI and is being silently dropped here) or delete the dead line outright. Same applies to the commented `<span>` placeholder around line 65. If the API-initiated phase column is intentionally removed, please mention it in the commit message / PR description — it's a user-visible regression that reviewers shouldn't have to discover from a diff.

File: .devcontainer/Dockerfile [hunk 5]
```diff
@@ -39,7 +39,7 @@ RUN apt-get update --fix-missing && \
 
 RUN apt-get install -y \
    gcc clang libclang-dev python3-pip python3-plumbum \
-   hub numactl openjdk-21-jre-headless maven
+   hub numactl openjdk-21-jdk maven
```
Comment (mythical-fred) [inline .devcontainer/Dockerfile:42]: Unrelated change in a "Checkpoint UI" PR — switching the devcontainer from `openjdk-21-jre-headless` to the full `openjdk-21-jdk` increases image size noticeably and has nothing to do with the web-console work. Either drop it or split into its own commit with a justification (why is the JDK now required?).

File: js-packages/web-console/src/lib/functions/common/date.ts [hunk 6]
```diff
@@ -1,3 +1,19 @@
...
+    return null
+  }
+  const hex = uuid.replace(/-/g, '').slice(0, 12)
+  const ms = parseInt(hex, 16)
+  if (!Number.isFinite(ms) || ms <= 0) {
+    return null
+  }
+  return Dayjs(ms)
+}
+
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/functions/common/date.ts:16]: Minor: the magic offsets (`uuid[14]`, `slice(0, 12)`) deserve a one-line comment naming what they correspond to (UUID version nibble, first 48 bits of timestamp). Also worth a unit test — this is a pure function and trivial to cover (valid v7 → expected ms; non-v7 → null; malformed input → null).

**Review by mythical-fred — CHANGES_REQUESTED** [review]
Three blockers, see inline. Tests for the new components are good (CheckpointDialog + CheckpointsStatus). Note: PR title says "Add checkpoint UI" but the commit also removes the existing API-initiated phase column from TransactionStatus — please call that out in the PR description or split it.

File: js-packages/web-console/src/lib/functions/common/date.ts (hunk 6 above)
Comment (mihaibudiu) [reply js-packages/web-console/src/lib/functions/common/date.ts:16]: I saw this before someplace

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointActivityStatus.svelte [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  const activity = $derived(metrics.current.checkpoint_activity)
+  const permanentErrors = $derived(metrics.current.permanent_checkpoint_errors)
+  const failure = $derived(checkpointStatus?.failure)
+  const lastSuccess = $derived(checkpointStatus?.success)
+
+  const showFailure = $derived(
+    failure != null && (lastSuccess == null || failure.sequence_number > lastSuccess)
+  )
+
+  const isPermanentlyUnavailable = $derived(
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointActivityStatus.svelte]: is there a way to avoid this repetition?

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    api.checkpointPipeline(pipelineName).finally(() => {
+      checkpointRequested = false
+    })
+  }
+
+  const openCheckpointDialog = () => {
+    globalDialog.dialog = checkpointDialog
+  }
+
+  const last = $derived(checkpoints.at(-1))
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte:50]: can this array be empty?

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+{#snippet checkpointDialog()}
+  <CheckpointDialog onConfirm={handleCheckpoint} />
+{/snippet}
+
+{#snippet allCheckpoints()}
+  {#if checkpoints.length > 0}
+    <div class="ml-auto flex items-center gap-4">
+      <span class=""
+        >{checkpoints.length}
+        {checkpoints.length === 1 ? 'checkpoint' : 'checkpoints'} at {humanSize(totalBytes)}</span
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte]: at?
maybe "using"?

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsStatus.svelte [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    {/if}
+  </div>
+  {#if onCheckpoint}
+    <button
+      class="btn w-full preset-outlined-primary-500 btn-sm"
+      data-testid="btn-make-checkpoint"
+      onclick={openCheckpointDialog}
+      disabled={requested}
+    >
+      {requested ? 'Requesting checkpoint...' : 'Make checkpoint'}
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsStatus.svelte]: "Make"?
"Create?"

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsStatus.svelte.spec.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
+import { describe, expect, it, vi } from 'vitest'
+import { page } from 'vitest/browser'
+import { render } from 'vitest-browser-svelte'
+import type { CheckpointMetadata } from '$lib/services/manager'
+
+const setDialogMock = vi.fn()
+
+vi.mock('$lib/compositions/layout/useGlobalDialog.svelte', () => ({
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsStatus.svelte.spec.ts:8]: isn't there a convention for naming test files?

**Review by mihaibudiu — APPROVED** [review]

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointActivityStatus.svelte (hunk 7 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointActivityStatus.svelte]: Good catch

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte (hunk 8 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsIndicator.svelte:50]: It can; .at() is a safe function that returns `undefined` if the array is empty, which I need to know

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsStatus.svelte.spec.ts (hunk 11 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/editor/performance/CheckpointsStatus.svelte.spec.ts:8]: Yes, there is; this block declares a mock for the specified module, imported by the component being tested

**Review by mythical-fred — COMMENTED** [review]
$No new commits since the re-review on `776bca5f` (https://github.com/feldera/feldera/pull/6065#pullrequestreview-4240761757). The five inline blockers from that review still stand:\n\n- `CheckpointsIndicator.svelte:76` — `|| true` always-true guard\n- `TabPerformance.svelte:20` — unused `usePremiumFeatures` import\n- `TransactionStatus.svelte:15,65` — silent removal of the api-initiated phase column\n- `.devcontainer/Dockerfile` — unrelated `jre-headless` → `jdk` swap\n- `date.ts uuidV7Timestamp` — magic offsets without a unit test\n\nHappy to re-review once those are addressed (and Mihai's open inline questions answered).

File: js-packages/web-console/src/lib/components/pipelines/editor/performance/TransactionStatus.svelte [hunk 12]
```diff
@@ -2,17 +2,16 @@
...
 
   const formatQty = (v: number) => format(',.0f')(v)
 
-  let { metrics }: { metrics: { current: PipelineMetrics } } = $props()
+  let { metrics, class: _class = '' }: { metrics: { current: PipelineMetrics }; class?: string } =
+    $props()
 
   const global = $derived(metrics.current.global)
   const transactionStatus = $derived(global.transaction_status)
   const commitProgress = $derived(global.commit_progress)
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/components/pipelines/editor/performance/TransactionStatus.svelte:14]: Still open from the prior review: the API-initiated phase column (`transaction_initiators.initiated_by_api` -> Started / Committed / None) was removed entirely in this PR and is not mentioned in the description. That's a user-visible regression in a checkpoint-UI PR. Either keep the column (perhaps relocated alongside the new progress widget) or call out the removal in the PR description and changelog so it's an intentional, documented change.

File: .devcontainer/Dockerfile (hunk 5 above)
Comment (mythical-fred) [inline .devcontainer/Dockerfile:42]: Still open from the prior review: `openjdk-21-jre-headless` -> `openjdk-21-jdk` in a web-console PR is unrelated to the checkpoint UI. Please drop this from this PR and submit it separately with its own justification (and a note about the image-size impact of pulling in the full JDK).

File: js-packages/web-console/src/lib/functions/common/date.ts [hunk 13]
```diff
@@ -1,3 +1,19 @@
...
+export function uuidV7Timestamp(uuid: string): Dayjs.Dayjs | null {
+  if (uuid[14] !== '7') {
+    return null
+  }
+  const hex = uuid.replace(/-/g, '').slice(0, 12)
+  const ms = parseInt(hex, 16)
+  if (!Number.isFinite(ms) || ms <= 0) {
+    return null
+  }
+  return Dayjs(ms)
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/functions/common/date.ts:14]: Still open from the prior review: the doc comment now explains the magic offsets, but `uuidV7Timestamp` is a pure parsing function and trivially testable -- no unit test was added. Please add a small spec covering: a known v7 UUID -> expected ms; a v4 UUID -> null; a malformed/short string -> null. Per the workspace rule, behavior change without unit/integration tests blocks the PR.

**Review by mythical-fred — CHANGES_REQUESTED** [review]
Three of the five blockers from the prior review are still open.

File: .devcontainer/Dockerfile (hunk 5 above)
Comment (mythical-fred) [inline .devcontainer/Dockerfile:42]: Still unrelated to the checkpoint UI work. Swapping `openjdk-21-jre-headless` for the full `openjdk-21-jdk` enlarges the devcontainer image with no rationale tied to this PR. Please revert here and submit as a separate PR with the reason (e.g., needed by a JDK-only tool).

File: js-packages/web-console/src/lib/functions/common/date.ts [hunk 14]
```diff
@@ -1,3 +1,19 @@
+import Dayjs from 'dayjs'
+
+/** Extracts the timestamp from a UUID v7 (first 48 bits = unix ms).
+ *  Returns null for non-v7 UUIDs. */
+export function uuidV7Timestamp(uuid: string): Dayjs.Dayjs | null {
```
Comment (mythical-fred) [inline js-packages/web-console/src/lib/functions/common/date.ts:5]: Still no unit test for `uuidV7Timestamp`. Please add `date.spec.ts` covering: a valid v7 UUID returns the expected ms (e.g., the canonical RFC 9562 examples), a non-v7 UUID returns `null`, an all-zero timestamp prefix returns `null`, malformed input returns `null`. Workspace rule: behavior changes ship with tests.

**Review by mythical-fred — CHANGES_REQUESTED** [review]
Re-review at d6a4cebe. The new tip is just the base branch catching up (rebase); no code changes since 25426021.

Of the 3 outstanding blockers from review 4246528974:

1. **TransactionStatus.svelte — API-initiated phase column removal.** Resolved. The PR description now explicitly notes: "Removed API-initiated transaction phase status information because it is generally not needed." Loss of that column is now disclosed.
2. **.devcontainer/Dockerfile — `jre-headless` → `jdk` swap.** Still present. Unrelated to checkpoint UI; please drop it from this PR or land it separately with justification.
3. **`date.ts` `uuidV7Timestamp` — no unit test.** Still missing. New behavior in `js-packages/web-console/src/lib/functions/common/date.ts` has no corresponding `date.spec.ts`. Hard rule: behavior change without tests blocks.

Cannot approve until 2 and 3 are addressed.
